src/i2c.rs

```diff
@@ -0,0 +1,287 @@
+
+                                unsafe {
+                                    w.ccr().bits(freq).duty().bit(duty)
+                                }
+                            });
+                        }
+                    };
+
+                    i2c.cr1.modify(|_, w| w.pe().set_bit());
+
+                    I2c { i2c, pins }
+                }
+
+                /// Releases the I2C peripheral and associated pins
+                pub fn free(self) -> ($I2CX, PINS) {
+                    (self.i2c, self.pins)
+                }
+            }
+
+            impl<PINS> Write for I2c<$I2CX, PINS> {
+                type Error = Error;
+
+                fn write(&mut self, addr: u8, bytes: &[u8]) -> Result<(), Error> {
+                    // TODO support transfers of more than 255 bytes
+                    assert!(bytes.len() < 256 && bytes.len() > 0);
+
+                    self.i2c.cr1.modify(|_, w| w.start().set_bit());
+                    busy_wait!(self.i2c, sb);
+
+                    self.i2c.dr.write(|w| unsafe { w.dr().bits(addr & 0b1111_1110) });
+                    busy_wait!(self.i2c, addr);
+                    let _ = self.i2c.sr2.read();
+
+                    for byte in bytes {
+                        busy_wait!(self.i2c, tx_e);
+                        self.i2c.dr.write(|w| unsafe { w.dr().bits(*byte) });
+                    }
+                    busy_wait!(self.i2c, tx_e);
+
+                    self.i2c.cr1.modify(|_, w| w.stop().set_bit());
+
+                    Ok(())
+                }
+            }
+
+            impl<PINS> WriteRead for I2c<$I2CX, PINS> {
+                type Error = Error;
+
+                fn write_read(
+                    &mut self,
+                    addr: u8,
+                    bytes: &[u8],
+                    buffer: &mut [u8],
+                ) -> Result<(), Error> {
+                    // TODO support transfers of more than 255 bytes
+                    assert!(bytes.len() < 256 && bytes.len() > 0);
+                    assert!(buffer.len() < 256 && buffer.len() > 0);
+
+                    self.i2c.cr1.modify(|_, w| w.start().set_bit());
+                    busy_wait!(self.i2c, sb);
+
+                    self.i2c.dr.write(|w| unsafe { w.dr().bits(addr & 0b1111_1110) });
+                    busy_wait!(self.i2c, addr);
+                    let _ = self.i2c.sr2.read();
+
+                    for byte in bytes {
+                        busy_wait!(self.i2c, tx_e);
+                        self.i2c.dr.write(|w| unsafe { w.dr().bits(*byte) });
+                    }
+                    busy_wait!(self.i2c, tx_e);
+
+                    self.i2c.cr1.modify(|_, w| w.start().set_bit());
+                    busy_wait!(self.i2c, sb);
+
+                    self.i2c.dr.write(|w| unsafe { w.dr().bits(addr | 0b0000_0001) });
+                    busy_wait!(self.i2c, addr);
+                    let _ = self.i2c.sr2.read();
+
+                    for byte in buffer {
```

Comment:
This looks like "method 1" in the manual, but "method 2" is supposed to work better for polling.

It also looks like it might be missing setting the `ACK` flag (it should be set before checking `rx_ne` for all the received bytes except the last).

Reply:
Thanks for the hint, gonna fix it.
I haven't yet tested read part at all, so it probably doesn't work.